### PR TITLE
Fix for: Pasting inside editor which contains a table after selecting all content is not working

### LINKED ISSUE #875: Pasting inside editor which contains a table after selecting all content is not working
## Are you reporting a feature request or a bug?

Bug

## Provide detailed reproduction steps (if any)

1. Open https://jsfiddle.net/9a6aenLg/2/
2. Select table via elements path and copy it.
3. Press <kbd>Ctrl</kbd> + <kbd>A</kbd>.
4. Paste table.

### Expected result

All content is replaced by pasted table.

### Actual result

Table inside editor's content becomes selected.

It seems like editor performs paste inside the table.

## Other details

* CKEditor version: 4.7.2
* Installed CKEditor plugins: tableselection

It's a regression from 4.7.1, probably due to the refactorization of paste algorithm.

### PR DESCRIPTION
**Pull request by Dumluregn**
## What is the purpose of this pull request?

Bugfix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What is the proposed changelog entry for this pull request?

```
[#875](https://github.com/ckeditor/ckeditor4/issues/875): Fix for: Pasting inside editor which contains a table with [Table Selection plugin](https://ckeditor.com/cke4/addon/tableselection) after selecting all content is not working.
```

## What changes did you make?

The issue here was caused by the custom `pasteListener()` for table selection plugin. It overrides the default paste logic by moving selection during paste to the first table cell and ignoring the rest of editor:
```javascript
// In case of mixed content or non table content just select first cell, and erase content of other selected cells.
// Selection is left in first cell, so that default CKEditor logic puts pasted content in the selection (#520).
if ( !pastedTable ) {
	selectCellContents( tableSel.cells.first );

	// Due to limitations of our undo manager, in case of mixed content
	// cells must be emptied after pasting (#520).
	editor.once( 'afterPaste', function() {
		editor.fire( 'lockSnapshot' );
		tableSel.emptyCells( tableSel.cells.all.slice( 1 ) );
		// Reselecting cells allows to create correct undo snapshot (#763).
		fakeSelectCells( editor, tableSel.cells.all );
		editor.fire( 'unlockSnapshot' );
	} );

	return;
}
```
To solve this, I added an extra condition for ignoring this custom pasting.

Closes #875.

### CODE REVIEW

File: plugins/tableselection/plugin.js [hunk 1]
```diff
@@ -811,9 +812,12 @@
...
 			// It's single range that does not fully contain table element and is not boundary, e.g. collapsed selection within
 			// cell, part of cell etc.
-			( selectedCells.length === 1 && !rangeContainsTableElement( selection.getRanges()[ 0 ] )  && !boundarySelection ) ||
+			( selectedCells.length === 1 && !rangeContainsTableElement( selection.getRanges()[ 0 ] ) && !boundarySelection ) ||
 			// It's a boundary position but with no table pasted.
-			( boundarySelection && !pastedTable ) ) {
+			( boundarySelection && !pastedTable ) ||
+			// There is some selected content before table (#875).
+			( selectionAncestor && selectionAncestor.$.nodeType == 1 && selectionAncestor.hasClass( 'cke_editable' ) )
+			) {
```
Comment (msamsel) [inline plugins/tableselection/plugin.js]: This `if` statement became too complicated and unreadable, please extract or simplify it. You can also combine this statement with check: https://github.com/ckeditor/ckeditor4/blob/ad381679b4436f489a65f6f397680da338024b0f/plugins/tableselection/plugin.js#L784-L786
and create a function like: `useTableSelection( selection )`
so entire code will  be reduce to:
```
if ( !useTableSelection( selection ) ) {
    return;
}
```
In `useTableSelection( selection )` you can also use the return early pattern to prevent unecessary checks.

File: plugins/tableselection/plugin.js [hunk 2]
```diff
@@ -811,9 +812,12 @@
 		if ( !selectedCells.length ||
 			// It's single range that does not fully contain table element and is not boundary, e.g. collapsed selection within
 			// cell, part of cell etc.
-			( selectedCells.length === 1 && !rangeContainsTableElement( selection.getRanges()[ 0 ] )  && !boundarySelection ) ||
+			( selectedCells.length === 1 && !rangeContainsTableElement( selection.getRanges()[ 0 ] ) && !boundarySelection ) ||
 			// It's a boundary position but with no table pasted.
-			( boundarySelection && !pastedTable ) ) {
+			( boundarySelection && !pastedTable ) ||
+			// There is some selected content before table (#875).
+			( selectionAncestor && selectionAncestor.$.nodeType == 1 && selectionAncestor.hasClass( 'cke_editable' ) )
```
Comment (msamsel) [inline plugins/tableselection/plugin.js]: Please don't make checks based on native element like this: `selectionAncestor.$.nodeType == 1`. There's appropriate abstraction for that, which should be for such purpose: https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_dom_element.html#property-type, https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR.html#property-NODE_ELEMENT

Comment (msamsel) [inline plugins/tableselection/plugin.js]: Checking `selectionAncestor.hasClass( 'cke_editable' )` is a bad idea. If you want to check if something is an editbale then you can use  [`equals`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editable.html#method-equals) method, for example:
```
editor.editable().equals( nodeToCheckWhichMightBeAnEditable );
```

**Comment by Dumluregn** [comment]
Short summary:
* made fix more generic (including mentioned `blockquote` issue);
* there is known bug in #541 which limits the possibilities of this fix;
* when selecting cell with a nested table it will still act the same way as it seems [impossibru](https://i.kym-cdn.com/entries/icons/original/000/004/918/imposibru.jpg) to fix without major refactor or breaking something else;
* refactored and improved code style;
* rebased onto lastest master.

I had some major doubts about explicitly naming table elements in [this line](https://github.com/ckeditor/ckeditor4/pull/3561/files#diff-3b0ed41f606e1508a1301dfcc588611eR876), but didn't manage to find any existing abstraction for them 🤔

File: plugins/tableselection/plugin.js [hunk 3]
```diff
@@ -777,53 +777,21 @@
...
-
-		// (#2945)
-		if ( table && table.hasAttribute( ignoredTableAttribute ) ) {
-			return;
-		}
-
-		var selectedCells = getSelectedCells( selection ),
+			selectedCells = getSelectedCells( selection ),
 			pastedTable = this.findTableInPastedContent( editor, evt.data.dataValue ),
 			boundarySelection = selection.isInTable( true ) && this.isBoundarySelection( selection ),
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tableselection/plugin.js]: Nice idea, and agree on `pastedTable`. But `boundarySelection` is necessary right after declarations in [`shouldUseCustomPaste()`](https://github.com/ckeditor/ckeditor4/pull/3561/files/b9e556ebf2a23f900ee6b40615f501ea6b2c79ff#diff-3b0ed41f606e1508a1301dfcc588611eR865) (to be renamed) and also [later in the listener itself](https://github.com/ckeditor/ckeditor4/pull/3561/files/b9e556ebf2a23f900ee6b40615f501ea6b2c79ff#diff-3b0ed41f606e1508a1301dfcc588611eR830), so I guess it should be left as is.

File: tests/plugins/tableselection/manual/integrations/clipboard/selectall.html [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+			</tr>
+		</tbody>
+	</table>
+
+	<p>After</p>
+	</blockquote>
+</div>
+
+<script>
+	if ( bender.tools.env.mobile ) {
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply tests/plugins/tableselection/manual/integrations/clipboard/selectall.html]: The issue is present there as well and can be reproduced so I unignored that test.

File: plugins/tableselection/plugin.js [hunk 5]
```diff
@@ -889,6 +857,56 @@
...
+				boundaryWithoutTable = boundarySelection && !pastedTable;
+
+			// (#2945)
+			if ( isIgnoredTable ) {
+				return false;
+			}
+
+			var selectionAncestor = selection.getCommonAncestor(),
+				selectionExceedesTable = selectionAncestor &&
+					selectionAncestor.is &&
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tableselection/plugin.js]: Yes, [getCommonAncestor()] can return [`CKEDITOR.dom.text`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_dom_text.html) (e.g. when you select text with a space between - like `So[me te]xt` ) which doesn't have `is()` method. I'll go with checking node type then. On the matter of the `getCommonAncestor()` method I think it would be better to just edit docs, but for now I'll just report issue and wait for decision.

File: plugins/tableselection/plugin.js [hunk 6]
```diff
@@ -889,6 +857,56 @@
...
+					!selectionAncestor.is( 'table', 'tbody', 'tr', 'td' );
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tableselection/plugin.js]: `tableContent` was what I was thinking about, but `rangeContainsTableElement()` works just as fine and it reduces the complexity of that conditional, so I will go with it, thanks. As for `selection.isInTable()` it's not viable because it returns 'true' while pasting inside single table cell.

**Comment by Dumluregn** [comment]
Using `rangeContainsTableElement()` seems to fix also the bug you mentioned [here](https://github.com/ckeditor/ckeditor4/pull/3561#pullrequestreview-319787118):
pastingInNestedTable

File: plugins/tableselection/plugin.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tableselection/plugin.js]: Well, I didn't notice at first but `pastedTable` is also used in both `isCustomPaste()` and in the listener, so it has to be calculated in the main scope right at the beginning.

**Comment by Dumluregn** [comment]
I had to add one positive early return in `isCustomPaste()` as tests pasting table into boundary selection had detected bugs.

File: plugins/tableselection/plugin.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (msamsel) [reply plugins/tableselection/plugin.js]: That's bad, it doesn't look like function requires those values. This might generate hard to spot errors in the future as a function became really tied to its context. I would go with direct passing those arguments to the function.

File: tests/plugins/tableselection/integrations/clipboard/pasteomittingtableselection.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+			}
+		}
+	};
+
+	var tests = {
+		setUp: function() {
+			bender.tools.ignoreUnsupportedEnvironment( 'tableselection' );
+		},
+
+		// // (#875)
```
Comment (msamsel) [inline tests/plugins/tableselection/integrations/clipboard/pasteomittingtableselection.js]:
```suggestion
		// (#875)
```

File: plugins/tableselection/plugin.js [hunk 8]
```diff
@@ -889,6 +858,76 @@
...
+
+			// 2. It's a boundary position but with no table pasted.
+			isBoundaryWithoutTable = ( boundarySelection && !pastedTable );
+
+			if ( isBoundaryWithoutTable ) {
+				return false;
+			}
+
+			// 2.1 If it's boundary position and table is pasted, proceed with custom paste.
+			if ( boundarySelection ) {
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tableselection/plugin.js]: I did that as I didn't see a reason why pasting results should be different with and without `tableselection`. But when I think about it for the second time it's rather obvious that the results of pasting with `ignoredTableAttribute` should be the same as without `tableselection` plugin installed, and to make it work the same as with `tableselection` (as I thought) it would require a big refactor and should be done in separate issue (if it should and could be done at all - I'm not sure it's even possible), so you're right `ignoreTableAttribute` should have a higher priority here.

File: plugins/tableselection/plugin.js [hunk 9]
```diff
@@ -889,6 +858,76 @@
...
+			// Do not customize paste process in following cases:
+			// 1. No cells are selected.
+			if ( !selectedCells.length ) {
+				return false;
+			}
+
+			// 2. It's a boundary position but with no table pasted.
+			isBoundaryWithoutTable = ( boundarySelection && !pastedTable );
+
+			if ( isBoundaryWithoutTable ) {
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tableselection/plugin.js]: Well, maybe I indeed created an information overflow here 🤔. As most of the conditions aren't that long, I'd give up creating variables and go with comments before `if`s as they have some added value in helping to understand the conditions, while variable names are sometimes unclear.

File: plugins/tableselection/plugin.js [hunk 10]
```diff
@@ -889,6 +858,76 @@
...
+			isIgnoredTable = table && table.hasAttribute( ignoredTableAttribute );
+
+			if ( isIgnoredTable ) {
+				return false;
+			}
+
+			// 5. It's single range that does not fully contain table element and is not boundary, e.g. collapsed selection within
+			// cell, part of cell etc.
+			isCollapsedPartial = selectedCells.length === 1 &&
+				!isTableElement &&
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tableselection/plugin.js]: On one hand it is totally logical, but on the other hand [here](https://github.com/ckeditor/ckeditor4/pull/3561#discussion_r351234420) we explicitly added unrequired parameters to function just to avoid misleading in the future. I'm afraid that if we remove conditions that are unnecessary with current check order here and [here](https://github.com/ckeditor/ckeditor4/pull/3561#discussion_r351240007), we will lose the information that they are still important and we don't place them just to keep code DRY. We could keep information about omitted conditions in comment, but it won't be save neither as code may desynchronize with comments as you mentioned [here](https://github.com/ckeditor/ckeditor4/pull/3561#discussion_r351238455). WDYT?

File: plugins/tableselection/plugin.js [hunk 11]
```diff
@@ -889,6 +858,76 @@
...
+				!boundarySelection;
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tableselection/plugin.js]: Please check out [this comment](https://github.com/ckeditor/ckeditor4/pull/3561#discussion_r351741962).

**Comment by Dumluregn** [comment]
Added mentioned tests:
* `test isBoundaryWithoutTable case` covers point 2. (btw now it's point 3.)
* `test isCollapsedPartial case` covers point 5.

**Comment by Dumluregn** [comment]
In the course of changing conditions for `inCustomPaste()` the last one called by me `isCollapsedPartial` became irrelevant as you noticed [here](https://github.com/ckeditor/ckeditor4/pull/3561#discussion_r351242426), so I removed it - everything seems to be working fine.

**Comment by msamsel** [comment]
I rebased it to current `master`.

File: plugins/tableselection/plugin.js [hunk 12]
```diff
@@ -889,6 +858,56 @@
...
+				return false;
+			}
+
+			// 3. It's a boundary position but with no table pasted.
+			if ( boundarySelection && !pastedTable ) {
+				return false;
+			}
+
+			// 3.1 If it's boundary position and table is pasted, proceed with custom paste.
+			if ( boundarySelection ) {
```
Comment (f1ames) [inline plugins/tableselection/plugin.js]: This looks a little confusing (all other conditions returns `false`) - I see it's intended (https://github.com/ckeditor/ckeditor4/pull/3561#issuecomment-557612461), however some more explicit comment why it needs such early return here could be very useful.

Apart from the explanation, I wonder if you couldn't merge it into last `if` like:

```js
if ( !boundarySelection && !rangeContainsTableElement( ranges[ 0 ] ) ) { ... }
```

File: tests/plugins/tableselection/manual/integrations/clipboard/selectall.md [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+	Selected content was replaced by pasted content.
+
+	### Unexpected:
+
+	Pasted data appeared in table, the rest of content is untouched.
+
+1. Undo changes.
+
+1. Create selection with content between brackets in table cell, paste.
```
Comment (f1ames) [inline tests/plugins/tableselection/manual/integrations/clipboard/selectall.md]: This step is a little unclear, I don't see any brackets in the table content (did you mean any of`([{` or something else here)?

**Comment by Dumluregn** [comment]
Rebased onto the latest master.

**Comment by f1ames** [comment]
Rebased onto latest `master`.
